src/type-coercions.md

```diff
@@ -182,5 +180,109 @@ unsized coercion to `Foo<U>`.
 > has been stabilized, the traits themselves are not yet stable and therefore
 > can't be used directly in stable Rust.
 
-[`Unsize`]: ../std/marker/trait.Unsize.html
+## Least upper bound coercions
+
+*Least upper bound coercion* (LUB coercion) is the form of type coercion take
+place when several expressions of possibly different types need be unified to a
+single type. This happens in [match] arms, [if expressions] and [array literal
+expressions]. In LUB coercion, the compiler tries to find the least upper
+bound of given types.
+
+For example:
+
+```rust
+# let (a, b, c) = (0, 1, 2);
+let foo = if true {
+    a
+} else if false {
+    b
+} else {
+    c
+};
+
+let bar = match 42 {
+    0 => a,
+    1 => b,
+    _ => c,
+};
+
+let baz = [a, b, c];
+```
+
+In this example, both `foo` and `bar` has the type
+`LubCoerce(typeof(a), typeof(a), typeof(c))` and `baz` has the type
+`[LubCoerce(typeof(a), typeof(b), typeof(c)); 3]`.
+
+LUB coercion is performed by the following algorithm:
+
+```txt
+Lub(a: Type, b: Type):
```

Comment:
Hmm, the actual code seems to have a bit more of a "bias", i.e., it has a notion of "old and new", and ordering I think maybe *can* matter. I wonder if we can write this section with a bit less specificity -- let me check how the coercion section is written.

Reply:
@nikomatsakis  Thanks for reviewing. Does the "old and new" means the `prev_ty` and `new_ty` in the code? I think it's represented in this part:
```
LubCoerce(vars...):
  result = vars.get(0)
  for var in vars[1..]:
    result = Lub(result, var)
  return result
```
Where `result` represents the `prev_ty` and `var` represents the `new_ty`. If you mean other things with "old and new". Could you elaborate more?

But for the ordering, I re-thinked about it and found it do matters in some cases(some edge cases). I will remove the property 1 later.